ando/AnDOChecker.py

```diff
@@ -0,0 +1,209 @@
+import os
+import os.path as op
+import re
+import argparse
+
+from AnDO.ando import rulesStructured as Rs
+
+
+def is_valid(input_directory):
+    """ Checks the ephys-BIDS validity of a data set.
+
+    The specifications that define what is checked by this function is available in the following document:
+
+
+    Parameters
+    ----------
+        input_directory : string
+            Name of the root directory containing the data set to be checked
+
+    Returns
+    -------
+        valid : boolean
+            True if the data set follows the ephys-BIDS specification; False if not
+
+        error_list : list
+            List of errors (empty if the data set is valid)
+
+    """
+
+    error_list = []
+
+    # walk through the directory tree using the os.walk function
+    for ind, (root, dirs, files) in enumerate(os.walk(input_directory)):
+        # compute depth of the directory given as input
+        if ind == 0:
+            # remove ending / or \ if the input directory was given with it at the end
+            if root.endswith(os.path.sep):
+                root = root[:-1]
+            # count the number of / or \ in the directory name to estimate its "depth"
+            initial_depth = root.count(os.path.sep)
+        # estimate at which level we are
+        depth = root.count(os.path.sep) - initial_depth
+        # print(depth,root,dirs,files)
+        ###
+        # extract rules for this level!
+        ###
+        currentdepth_rules = Rs.rules_set[depth]
+        ###
+        # 1.check whether the "mandatory Folders" are present at this level
+        #
+        ###
+        if len(currentdepth_rules['mandatory_folders']) > 0:
+            # loop over rules, each rule corresponding to one mandatory folder
+            for current_mandatoryfolder_rule in (currentdepth_rules['mandatory_folders']):
+                # create the list of regexp
+                list_of_mandatory_folders = (
+                    get_list_of_rules_with(current_mandatoryfolder_rule))
+                for each_file in list_of_mandatory_folders:
+                    dir_res = [re.compile(each_file).search(d) is None for d in dirs]
+                    if all(dir_res):
+                        error_list.append(
+                            "Mandatory folder not found for this rule : {}".format(current_mandatoryfolder_rule))
+        ###
+        # 2. check whether the rules are followed for the folder at this level ("authorized folders")
+        #
+        ###
+        folder = op.split(root)[1]
+        folder_res = [re.compile(x).search(folder) is None
```

Comment:
This kind of list comprehension is repeatedly used in the code. Should this be separated into a utility function, so it can also be documented better?

Reply:
Done